src/somnium/congomongo.clj

```diff
@@ -95,20 +97,39 @@
   [^String host ^Integer port]
   (ServerAddress. host port))
 
+(defn- make-mongo-client
+  ([addresses creds options]
+    (if (> (count addresses) 1)
+      (MongoClient. ^java.util.List addresses creds options)
+      (MongoClient. ^ServerAddress (first addresses) creds options)))
+
+  ([addresses options]
+    (if (> (count addresses) 1)
+      (MongoClient. ^java.util.List addresses options)
+      (MongoClient. ^ServerAddress (first addresses) options))))
+
 (defn- make-connection-args
   "Makes a connection with passed database name, host, port and MongoClientOptions"
-  [db args]
-    (let [instances (take-while #(not (instance? MongoClientOptions %)) args)
-          addresses (->> (if (keyword? (first instances))
-                           (list (apply array-map instances)) ; Handle legacy connect args
-                           instances)
-                      (map (fn [{:keys [host port] :or {host "127.0.0.1" port 27017}}]
-                             (make-server-address host port))))
-          ^MongoClientOptions options (or (first (filter #(instance? MongoClientOptions %) args)) (mongo-options))
-          mongo (if (> (count addresses) 1)
-                  (MongoClient. ^java.util.List addresses options)
-                  (MongoClient. ^ServerAddress (first addresses) options))
-          n-db (if db (.getDB mongo db) nil)]
+  [db {:keys [instances options username password]}]
+  (when (or username password)
+    (assert (and username password) "Username and password must be supplied for authenticated connections"))
```

Comment:
I could have it throw an `IllegalArgumentException` either?

I'd rather throw a specific exception with a helpful message than NPE on the password to `char[]` conversion or whatever Mongo throws if the username is `nil` later down the call stack.